### PR TITLE
Add QueryServiceAvailable in ShowCollections

### PR DESCRIPTION
**Pull request by letian-jiang**
Add QueryServiceAvailable in QueryCoord ShowCollections, indicateing whether the collection has any available replica to serve query requests

Related to #17098 

Signed-off-by: Letian Jiang <letian.jiang@zilliz.com>

### CODE REVIEW

File: internal/querycoord/impl.go [hunk 1]
```diff
@@ -88,6 +88,58 @@ func (qc *QueryCoord) GetStatisticsChannel(ctx context.Context) (*milvuspb.Strin
...
+				shards[shard.DmChannelName] = list
+			}
+		}
+	}
+
+	shardLeaderLists := make([]*querypb.ShardLeadersList, 0, len(shards))
+	for _, shard := range shards {
+		shardLeaderLists = append(shardLeaderLists, shard)
+	}
+	return len(shardLeaderLists) > 0
```
Comment (congqixia) [inline internal/querycoord/impl.go:140]: `return len(shards) > 0` will do.

File: internal/querycoord/impl.go [hunk 2]
```diff
@@ -88,6 +88,58 @@ func (qc *QueryCoord) GetStatisticsChannel(ctx context.Context) (*milvuspb.Strin
...
+						break
+					}
+				}
+			}
+
+			if isShardAvailable {
+				list.NodeIds = append(list.NodeIds, shard.LeaderID)
+				list.NodeAddrs = append(list.NodeAddrs, shard.LeaderAddr)
+				shards[shard.DmChannelName] = list
+			}
```
Comment (congqixia) [inline internal/querycoord/impl.go:132]: As previously discussed, this part checks whether one shard can provided service. What we want shall be whether each shard has at least one replica could provide query service.

Comment (letian-jiang) [reply internal/querycoord/impl.go:132]: Gonna fix it later.

Comment (letian-jiang) [reply internal/querycoord/impl.go:136]: Done

File: internal/querycoord/impl.go [hunk 3]
```diff
@@ -88,6 +88,63 @@ func (qc *QueryCoord) GetStatisticsChannel(ctx context.Context) (*milvuspb.Strin
 	}, nil
 }
 
+func (qc *QueryCoord) checkAnyReplicaAvailable(collectionID UniqueID) bool {
+	replicas, _ := qc.meta.getReplicasByCollectionID(collectionID)
+	// Note: copy from GetShardLeaders
+	numShards := 2
+	if len(replicas) > 0 {
+		numShards = len(replicas[0].ShardReplicas)
+	}
```
Comment (congqixia) [inline internal/querycoord/impl.go:97]: NIT: Feels weird that it has to have replicas to get shard number right. Maybe shard number could be retrieved from collection meta.

File: internal/querycoord/impl.go [hunk 4]
```diff
@@ -88,6 +88,63 @@ func (qc *QueryCoord) GetStatisticsChannel(ctx context.Context) (*milvuspb.Strin
...
+	shards := make(map[string]*querypb.ShardLeadersList)
+	shardNodes := getShardNodes(collectionID, qc.meta)
+	for _, replica := range replicas {
+		for _, shard := range replica.ShardReplicas {
+			list, ok := shards[shard.DmChannelName]
+			if !ok {
+				list = &querypb.ShardLeadersList{
+					ChannelName: shard.DmChannelName,
+					NodeIds:     make([]int64, 0),
+					NodeAddrs:   make([]string, 0),
```
Comment (congqixia) [inline internal/querycoord/impl.go:107]: NIT: make it nil will do the trick

File: internal/querycoord/impl.go [hunk 5]
```diff
@@ -1107,6 +1166,11 @@ func (qc *QueryCoord) GetShardLeaders(ctx context.Context, req *querypb.GetShard
 	}
 
 	replicas, err := qc.meta.getReplicasByCollectionID(req.CollectionID)
+	numShards := 2
+	if len(replicas) > 0 {
+		numShards = len(replicas[0].ShardReplicas)
+	}
```
Comment (congqixia) [inline internal/querycoord/impl.go:1172]: Same here

File: internal/querycoord/impl.go (hunk 4 above)
Comment (letian-jiang) [reply internal/querycoord/impl.go]: Fixed.

**Review by congqixia — APPROVED** [review]
/lgtm
please add ut for `checkAnyReplicaAvailable` later

**Pull request merged**